homeassistant/components/history/__init__.py

```diff
@@ -300,13 +362,409 @@ def _sorted_significant_states_json(
+            )
+        ):
+            _async_send_empty_response(connection, msg_id, start_time, end_time)
+            return
+
+        connection.subscriptions[msg_id] = callback(lambda: None)
+        connection.send_result(msg_id)
+        await _async_send_historical_states(
+            hass,
+            connection,
+            msg_id,
+            start_time,
+            end_time,
+            entity_ids,
+            filters,
+            include_start_time_state,
+            significant_changes_only,
+            minimal_response,
+            no_attributes,
+            True,
+        )
+        return
+
+    subscriptions: list[CALLBACK_TYPE] = []
+    stream_queue: asyncio.Queue[Event] = asyncio.Queue(MAX_PENDING_HISTORY_STATES)
+    live_stream = HistoryLiveStream(
+        subscriptions=subscriptions, stream_queue=stream_queue
+    )
+
+    @callback
+    def _unsub(*_utc_time: Any) -> None:
+        """Unsubscribe from all events."""
+        for subscription in subscriptions:
+            subscription()
+        subscriptions.clear()
+        if live_stream.task:
+            live_stream.task.cancel()
+        if live_stream.wait_sync_task:
+            live_stream.wait_sync_task.cancel()
+        if live_stream.end_time_unsub:
+            live_stream.end_time_unsub()
+            live_stream.end_time_unsub = None
+
+    if end_time:
+        live_stream.end_time_unsub = async_track_point_in_utc_time(
+            hass, _unsub, end_time
+        )
+
+    @callback
+    def _queue_or_cancel(event: Event) -> None:
+        """Queue an event to be processed or cancel."""
+        try:
+            stream_queue.put_nowait(event)
+        except asyncio.QueueFull:
+            _LOGGER.debug(
+                "Client exceeded max pending messages of %s",
+                MAX_PENDING_HISTORY_STATES,
+            )
+            _unsub()
+
+    _async_subscribe_events(
+        hass,
+        subscriptions,
+        _queue_or_cancel,
+        entities_filter,
+        entity_ids,
+        significant_changes_only=significant_changes_only,
+        minimal_response=minimal_response,
+    )
+    subscriptions_setup_complete_time = dt_util.utcnow()
+    connection.subscriptions[msg_id] = _unsub
+    connection.send_result(msg_id)
+    # Fetch everything from history
+    last_event_time = await _async_send_historical_states(
+        hass,
+        connection,
+        msg_id,
+        start_time,
+        subscriptions_setup_complete_time,
+        entity_ids,
+        filters,
+        include_start_time_state,
+        significant_changes_only,
+        minimal_response,
+        no_attributes,
+        True,
+    )
+
+    live_stream.task = asyncio.create_task(
+        _async_events_consumer(
+            subscriptions_setup_complete_time,
+            connection,
+            msg_id,
+            stream_queue,
+            no_attributes,
+        )
+    )
+
+    if msg_id not in connection.subscriptions:
```

Comment:
We can move this above the creating the `_async_events_consumer` task and avoid it even being created

Reply:
191353df1f53637da3ab587bfb52771932be708d